### ISSUE TITLE
actionlint detect warning when using inputs type conversion with fromJSON

### ISSUE THREAD

**Issue by guitarrapc** [rhysd/actionlint#181]
## Summary

actionlint warn when trying to use JSON data type conversion pattern and it's too strict, or other word too smart then what agent do.

Warning message

```bash
$ actionlint

.github\workflows\actionlint_issue.yaml:15:30: 1st argument of function call is not assignable. "bool" cannot be assigned to "string". called function type is "fromJSON(string) -> any" [expression]
   |
15 |       is-valid: ${{ fromJson(github.event.inputs.is-valid) }}
```

## Detail

When reusable workflow have inputs type `boolean` and want pass boolean from workflow_dispatch, JSON data type convert with `fromJSON` expression is officially introduced pattern. 

> see: https://docs.github.com/en/actions/learn-github-actions/expressions#example-returning-a-json-data-type

Without `fromJSON`, `github.event.inputs.foo` treat as `string` and cannot pass to workflow_call.
However actionlint warn this and seems guessing parameter as boolean is too smart then what agent do.

## Reproduce code

1. Prepare 2 workflows `actionlint.yaml` and `_reusable.yaml`

```yaml
name: actionlint issue

on:
  workflow_dispatch:
    inputs:
      is-valid:
        required: true
        description: "is-valid: true or false"
        type: boolean

jobs:
  issue:
    uses: ./.github/workflows/_reusable.yaml
    with:
      is-valid: ${{ fromJson(github.event.inputs.is-valid) }}
```

```yaml
name: _reusable

on:
  workflow_call:
    inputs:
      is-valid:
        required: true
        description: username to show
        type: boolean

jobs:
  reusable_workflow_job:
    runs-on: ubuntu-latest
    steps:
      - name: called is-valid
        run: echo "called is-valid. ${{ inputs.is-valid }}"
```

2. run actionlint.

## Workaround

Pass string to `fromJSON` not boolean, via `format('{0}', github.event.inputs.foo)`.

```yaml
name: actionlint issue

<reducted>

jobs:
  issue:
    uses: ./.github/workflows/_reusable.yaml
    with:
      is-valid: ${{ fromJson(format('{0}', github.event.inputs.is-valid)) }}
```

**Comment by rhysd** [comment]
Thank you for the detailed report with reproduction. This looks a bug of actionlint.

**Comment by rhysd** [comment]
Type of the input is specified as `boolean` at `on.workflow_dispatch.inputs.<name>.type` but it is only used for validation. Type of `inputs.<name>` is always string. But actionlint statically types `inputs.<name>` is `boolean` looking at the specification.

**Comment by rhysd** [comment]
I made a question at https://github.com/orgs/community/discussions/29796

**Comment by rhysd** [comment]
@guitarrapc By the way, `inputs` context instead of `github.event.inputs` solves this issue. It is typed correctly as boolean. If you still have this issue, please try the following:

```
is-valid: ${{ inputs.is-valid }}
```

**Comment by johnfoconnor** [comment]
Im running into this issue with the following code

`Error: .github/workflows/populate_caches.yaml:38:16: input "timeout" is typed as number by reusable workflow "./.github/workflows/unity-tests.yaml". string value cannot be assigned [expression]`


```
    strategy:
      matrix:
        timeout:
          # hack to allow a configurable number timeout as input. see
          # https://docs.github.com/en/actions/learn-github-actions/expressions#example-returning-a-json-data-type
          - ${{ fromJSON(inputs.timeout || '40') }}
    with:
      timeout: ${{ matrix.timeout }}
```
